src/result/conversion.rs

```diff
@@ -0,0 +1,119 @@
+use crate::{Bytes, Decimal, Int, IonType, Sequence, Str, Struct, Symbol, Timestamp};
+
+use std::marker::PhantomData;
+use thiserror::Error;
+
+pub type ConversionOperationResult<FromType, ToType> =
+    Result<ToType, ConversionOperationError<FromType, ToType>>;
+pub type ConversionResult<ToType> = Result<ToType, ConversionError>;
+
+/// Represents a mismatch during conversion between the expected type and the actual value's type.
+#[derive(Clone, Debug, Error, PartialEq)]
+#[error("Type conversion error; expected a(n) {output_type}, found a(n) {input_type}")]
+pub struct ConversionError {
+    input_type: String,
+    output_type: String,
+}
+
+impl<FromType, ToType> From<ConversionOperationError<FromType, ToType>> for ConversionError
+where
+    FromType: ValueTypeExpectation,
+    ToType: TypeExpectation,
+{
+    fn from(err: ConversionOperationError<FromType, ToType>) -> Self {
+        ConversionError {
+            input_type: err.input_value.expected_type_name(),
+            output_type: ToType::expected_type_name(),
+        }
+    }
+}
+
+/// Represents a mismatch during conversion between the expected type and the actual value's type;
+/// Holds the original value that was not able to be converted.
+#[derive(Clone, Debug, Error, PartialEq)]
+#[error("Type conversion error; expected a(n) {}, found a(n) {}",
+            ToType::expected_type_name(), .input_value.expected_type_name())]
+pub struct ConversionOperationError<FromType, ToType>
+where
+    FromType: ValueTypeExpectation,
+    ToType: TypeExpectation,
+{
+    input_value: Box<FromType>,
```

Comment:
Why does this need to be boxed? Are you concerned about stack size?

Reply:
Without boxing, clippy warns:

```shell
warning: the `Err`-variant returned from this function is very large
   --> src/element/mod.rs:492:34
    |
492 |     pub fn try_into_int(self) -> ConversionOperationResult<Element, Int> {
    |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the `Err`-variant is at least 128 bytes
    |
    = help: try reducing the size of `result::conversion::ConversionOperationError<element::Element, types::integer::Int>`, for example by boxing large elements or replacing it with `Box<result::conversion::ConversionOperationError<element::Element, types::integer::Int>>`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#result_large_err
    = note: `#[warn(clippy::result_large_err)]` on by default
```